cli/src/main/java/de/jplag/cli/logger/TerminalPrinter.java

```diff
@@ -0,0 +1,87 @@
+package de.jplag.cli.logger;
+
+import java.io.PrintStream;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+/**
+ * Prints strings to stdout. Provides the option to delay the actual printing.
+ */
+public class TerminalPrinter {
+    private static TerminalPrinter instance;
+
+    /**
+     * Threadsafe singleton getter
+     * @return The singleton instance
+     */
+    public static TerminalPrinter getInstance() {
+        synchronized (TerminalPrinter.class) {
+            if (instance == null) {
+                synchronized (TerminalPrinter.class) {
+                    instance = new TerminalPrinter();
+                }
+            }
+        }
+
+        return instance;
+    }
+
+    private final Queue<String> outputQueue;
+    private PrintStream targetStream;
+
+    private boolean isDelayed;
+
+    private TerminalPrinter() {
+        this.outputQueue = new PriorityQueue<>();
+        this.targetStream = System.out;
+        this.isDelayed = false;
+    }
+
+    /**
+     * Prints the given string to the terminal appending a line-break
+     * @param output The string to print
+     */
+    public void println(String output) {
```

Comment:
If you are synchronizing the whole method body anyways and the class is the monitor, just add `synchronized` as method keyword.